### PR TITLE
Mapping query builder

### PR DESCRIPTION
**Pull request by dantleech**
Replaces #1272.

Integrates the new document manager features currently in `develop`.

- Metadata mapping: The subscribers that were previously simply mapping values now only have to provide some metadata (which will eventually be cached -- so they won't even be called in the future).
- Query Builder: Adds DI configuration for the query builder.

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Document/BasePageDocument.php [hunk 1]
```diff
@@ -154,6 +154,8 @@ class BasePageDocument implements
      */
     protected $locale;
 
+    protected $originalLocale;
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Document/BasePageDocument.php]: What is this? If it is required then there should be a doc comment.

File: src/Sulu/Bundle/ContentBundle/Document/BasePageDocument.php [hunk 2]
```diff
@@ -305,6 +307,9 @@ public function getRedirectType()
      */
     public function setRedirectType($redirectType)
     {
+        if ($redirectType === null) {
+            die('asd');
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Document/BasePageDocument.php]: Destroy this.

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
+<?php
+
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: header

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder;
+use Sulu\Component\DocumentManager\DocumentStrategyInterface;
+use Sulu\Component\Content\Metadata\Factory\StructureMetadataFactory;
+
+class StructureQueryBuilderConverter extends QueryBuilderConverter
+{
+    private $structureMetaFactory;
+    private $structureMap = array();
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: docs

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        $this->structureMap = $queryBuilder->getStructureMap();
+        return parent::getQuery($queryBuilder);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function getPhpcrProperty($alias, $field)
+    {
+        if (false !== $position = strpos($field, 'structure#')) {
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Invert this IF statement.

File: src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php [hunk 6]
```diff
@@ -46,52 +35,45 @@ public static function getSubscribedEvents()
...
-        if ($node->hasProperty($propertyName)) {
+        // TODO: This does not seem quite right..
+        if ($document->getSuluOrder()) {
             return;
         }
 
+        $node = $event->getNode();
         $parent = $node->getParent();
         $nodeCount = count($parent->getNodes());
         $order = ($nodeCount + 1) * 10;
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: @danrot @wachterjohannes either I dont understand how this works, or it doesnt work (`sulu:order`). It seems that it only sets the order if it hasn't yet been set. I won't fix it in this PR, but FYI if you notice something.

File: src/Sulu/Component/Content/Document/Subscriber/ResourceSegmentSubscriber.php [hunk 7]
```diff
@@ -20,6 +20,10 @@
 use Sulu\Component\DocumentManager\Events;
 use Sulu\Component\DocumentManager\Event\AbstractMappingEvent;
 
+/**
+ * TODO: This could be made into a pure metadata subscriber if we make
+ *       the resource locator a system property.
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/ResourceSegmentSubscriber.php]: i.e. if we integrate RoutingAuto.

File: tests/Sulu/Component/Content/Document/Query/QueryBuilderTest.php [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Sulu\Component\Content\Document\Query;
```
Comment (dantleech) [inline tests/Sulu/Component/Content/Document/Query/QueryBuilderTest.php]: header

**Comment by dantleech** [comment]
Rebased and reviewed, need to do the above fixes but seems OK. Feel free to review @danrot @wachterjohannes 

We also need a new tag of PHPCR-ODM either before merging or before the next Sulu release, I created: https://github.com/doctrine/phpcr-odm/issues/645

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+namespace Sulu\Component\Content\Document\Query;
+
+use Doctrine\ODM\PHPCR\Query\Builder\QueryBuilder as BaseQueryBuilder;
+
+class QueryBuilder extends BaseQueryBuilder
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+use Doctrine\ODM\PHPCR\Query\Builder\QueryBuilder as BaseQueryBuilder;
+use Sulu\Component\Content\Document\Query\QueryBuilder;
+use PHPCR\SessionInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder;
+use Sulu\Component\DocumentManager\DocumentStrategyInterface;
+use Sulu\Component\Content\Metadata\Factory\StructureMetadataFactory;
+
+class StructureQueryBuilderConverter extends QueryBuilderConverter
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        MetadataFactoryInterface $metadataFactory,
+        PropertyEncoder $encoder,
+        DocumentStrategyInterface $strategy,
+        StructureMetadataFactory $structureMetaFactory
+    ) {
+        parent::__construct($session, $eventDispatcher, $metadataFactory, $encoder, $strategy);
+        $this->structureMetaFactory = $structureMetaFactory;
+    }
+
+    public function getQuery(BaseQueryBuilder $queryBuilder)
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php (hunk 6 above)
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: What exactly should this subscriber do? But you are right, it also feels weird to me.

File: src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php [hunk 12]
```diff
@@ -46,52 +35,45 @@ public static function getSubscribedEvents()
...
 
-        $node->setProperty($propertyName, $order, PropertyType::LONG);
-        $this->handleHydrate($event);
-    }
-
-    /**
-     * @param HydrateEvent $event
-     */
-    public function handleHydrate(AbstractMappingEvent $event)
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: Why can `handleHydrate` be removed.

File: src/Sulu/Component/Content/Document/Subscriber/RouteSubscriber.php [hunk 13]
```diff
@@ -10,80 +10,38 @@
...
-        }
-
-        $document->setTargetDocument($targetDocument);
     }
 
-    /**
-     * @param PersistEvent $event
-     */
-    public function doPersist(PersistEvent $event)
+    public function handleMetadataLoad(MetadataLoadEvent $event)
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/RouteSubscriber.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Subscriber/ShadowLocaleSubscriber.php [hunk 14]
```diff
@@ -68,6 +70,27 @@ public static function getSubscribedEvents()
         );
     }
 
+    public function handleMetadataLoad(MetadataLoadEvent $event)
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/ShadowLocaleSubscriber.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Subscriber/WebspaceSubscriber.php [hunk 15]
```diff
@@ -63,11 +62,7 @@ public function doHydrate(AbstractMappingEvent $event)
         $event->getAccessor()->set('webspaceName', $webspaceName);
     }
 
-    /**
-     * @param PersistEvent $event
-     */
     public function doPersist(PersistEvent $event)
     {
-        $this->doHydrate($event);
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/WebspaceSubscriber.php]: Shouldn't this method be removed as well?

File: src/Sulu/Component/Content/Document/Subscriber/WorkflowStageSubscriber.php [hunk 16]
```diff
@@ -20,78 +20,82 @@
...
-        $document = $event->getDocument();
+        return array(
+            Events::METADATA_LOAD => 'handleMetadataLoad',
+            Events::PERSIST => 'handlePersist',
+        );
+    }
 
-        $workflowStage = $this->getWorkflowStage($node, $locale);
-        $document->setWorkflowStage($workflowStage);
+    public function handleMetadataLoad(MetadataLoadEvent $event)
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/WorkflowStageSubscriber.php]: PHPDoc

File: src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php (hunk 6 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: I think it should update the sulu:order property, which I suspect is
now only set when creating new documents.

File: src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php (hunk 12 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: If I understand the context here, the removal of the hydrate/persist
methods where possible is the point of this PR.

By adding the metadata mapping hydrate and persist can be handled in the
"core" (and consequently, if we had a metadata cache, this subscriber
would not even be called).

Comment (danrot) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: Can you add a link to the class doing this now?

Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/OrderSubscriber.php]: https://github.com/sulu-io/sulu-document-manager/blob/develop/lib/Subscriber/Core/MappingSubscriber.php

**Comment by danrot** [comment]
Is this PR still valid? Or do you have to make a new one?

**Comment by dantleech** [comment]
I will adapt this PR.

**Comment by dantleech** [comment]
Replaced by #1863

**Pull request closed without merging**